File: ur10e_teleop_real_py/src/dashboard_client.py
"""
dashboard_client.py — minimal client for UR Dashboard Server (port 29999).

Used to automate the Power On → Booting → Release Brakes sequence at
startup and the reverse (Brake Engage → Power Off) at shutdown, so the
operator doesn't have to step through the pendant by hand.

Reference: https://www.universal-robots.com/articles/ur/dashboard-server-e-series-port-29999/
"""

import socket
import time


# Robot mode enum (e-Series)
ROBOT_MODE_NO_CONTROLLER   = 0
ROBOT_MODE_DISCONNECTED    = 1
ROBOT_MODE_CONFIRM_SAFETY  = 2
ROBOT_MODE_BOOTING         = 3
ROBOT_MODE_POWER_OFF       = 4
ROBOT_MODE_POWER_ON        = 5
ROBOT_MODE_IDLE            = 6
ROBOT_MODE_RUNNING         = 7

_ROBOT_MODE_NAMES = {
    0: 'NO_CONTROLLER', 1: 'DISCONNECTED', 2: 'CONFIRM_SAFETY',
    3: 'BOOTING', 4: 'POWER_OFF', 5: 'POWER_ON', 6: 'IDLE', 7: 'RUNNING',
}


def robot_mode_name(m):
    return _ROBOT_MODE_NAMES.get(m, f'?{m}?')


class DashboardClient:
    """Blocking TCP client for port 29999. Not threadsafe; use from one thread."""

    DEFAULT_PORT = 29999

    def __init__(self, host, port=DEFAULT_PORT, timeout=5.0, logger=None):
        self.host = host
        self.port = port
        self.timeout = timeout
        self.log = logger or (lambda s: print(f'[dashboard] {s}'))
        self._sock = None

    # ---- connection ---------------------------------------------------------

    def connect(self):
        if self._sock is not None:
            return True
        try:
            s = socket.create_connection((self.host, self.port), timeout=self.timeout)
            s.settimeout(self.timeout)
            # welcome banner — discard
            s.recv(4096)
            self._sock = s
            return True
        except Exception as e:
            self.log(f'connect({self.host}:{self.port}) failed: {e}')
            self._sock = None
            return False

    def close(self):
        if self._sock is not None:
            try:
                self._sock.close()
            except Exception:
                pass
            self._sock = None

    # ---- primitive ----------------------------------------------------------

    def _ask(self, cmd):
        """Send one command line and return the single-line response."""
        if self._sock is None and not self.connect():
            return ''
        try:
            self._sock.sendall((cmd + '\n').encode('utf-8'))
            data = b''
            while b'\n' not in data:
                chunk = self._sock.recv(4096)
                if not chunk:
                    break
                data += chunk
            return data.decode('utf-8', errors='replace').strip()
        except Exception as e:
            self.log(f'_ask({cmd!r}) failed: {e}')
            self.close()
            return ''

    # ---- state queries ------------------------------------------------------

    def get_robot_mode(self):
        """Return integer robot mode (see ROBOT_MODE_* constants)."""
        resp = self._ask('robotmode')
        # response: "Robotmode: RUNNING"
        name = resp.rsplit(':', 1)[-1].strip() if ':' in resp else resp.strip()
        name_to_int = {v: k for k, v in _ROBOT_MODE_NAMES.items()}
        return name_to_int.get(name, -1)

    def get_safety_mode(self):
        return self._ask('safetymode')

    def is_in_remote_control(self):
        return self._ask('is in remote control').strip().lower().startswith('true')

    # ---- power + brakes -----------------------------------------------------

    def power_on(self):
        return self._ask('power on')

    def power_off(self):
        return self._ask('power off')

    def brake_release(self):
        return self._ask('brake release')

    def stop_program(self):
        return self._ask('stop')

    def unlock_protective_stop(self):
        return self._ask('unlock protective stop')

    # ---- helpers ------------------------------------------------------------

    def wait_for_mode(self, target_mode, timeout=30.0, poll_interval=0.25):
        """Poll robotmode until it reaches target_mode or timeout. Returns bool."""
        deadline = time.time() + timeout
        last = -1
        while time.time() < deadline:
            m = self.get_robot_mode()
            if m != last:
                self.log(f'robot_mode = {m} ({robot_mode_name(m)})')
                last = m
            if m == target_mode:
                return True
            time.sleep(poll_interval)
        self.log(f'wait_for_mode({target_mode}) timed out '
                 f'(last seen {last} = {robot_mode_name(last)})')
        return False

    # ---- high-level sequences ----------------------------------------------

    def power_up_sequence(self, timeout=45.0):
        """Power On → (BOOTING →) IDLE → brake release → RUNNING."""
        if not self.connect():
            return False

        if not self.is_in_remote_control():
            self.log('WARNING: robot is NOT in Remote Control mode — '
                     'power-cycle will not work remotely.')

        mode = self.get_robot_mode()
        self.log(f'current robot_mode = {mode} ({robot_mode_name(mode)})')

        if mode == ROBOT_MODE_RUNNING:
            self.log('already RUNNING — power-up sequence skipped')
            return True

        self.log('power on …')
        self.power_on()
        if not self.wait_for_mode(ROBOT_MODE_IDLE, timeout=timeout):
            return False

        self.log('brake release …')
        self.brake_release()
        if not self.wait_for_mode(ROBOT_MODE_RUNNING, timeout=timeout):
            return False

        self.log('power-up sequence complete (RUNNING)')
        return True

    def power_down_sequence(self, timeout=30.0):
        """Stop program → power off. (UR engages brakes on power-off.)"""
        if not self.connect():
            return False
        self.log('stop program …')
        self.stop_program()
        time.sleep(0.5)
        self.log('power off …')
        self.power_off()
        self.wait_for_mode(ROBOT_MODE_POWER_OFF, timeout=timeout)
        self.log('power-down sequence complete')
        return True

File: ur10e_teleop_control_ff_cpp/src/dashboard_client.cpp
// dashboard_client.cpp — raw-socket client for UR Dashboard (port 29999).

#include "ur10e_teleop_control_ff_cpp/dashboard_client.hpp"

#include <algorithm>
#include <chrono>
#include <cstdio>
#include <cstring>
#include <thread>

#include <arpa/inet.h>
#include <netdb.h>
#include <sys/socket.h>
#include <sys/types.h>
#include <unistd.h>
#include <fcntl.h>

namespace ur10e_teleop_control_ff_cpp {

namespace {
const char* MODE_NAMES[] = {
    "NO_CONTROLLER", "DISCONNECTED", "CONFIRM_SAFETY", "BOOTING",
    "POWER_OFF", "POWER_ON", "IDLE", "RUNNING"
};
}  // namespace

const char* robot_mode_name(int m) {
  if (m < 0 || m > 7) return "?";
  return MODE_NAMES[m];
}

DashboardClient::DashboardClient(const std::string& host, int port,
                                  double timeout_sec)
    : host_(host), port_(port), timeout_sec_(timeout_sec) {}

DashboardClient::~DashboardClient() { close(); }

bool DashboardClient::connect() {
  if (sock_fd_ >= 0) return true;
  int fd = ::socket(AF_INET, SOCK_STREAM, 0);
  if (fd < 0) {
    std::fprintf(stderr, "[dashboard] socket() failed: %s\n", std::strerror(errno));
    return false;
  }

  struct timeval tv;
  tv.tv_sec  = static_cast<time_t>(timeout_sec_);
  tv.tv_usec = static_cast<suseconds_t>((timeout_sec_ - tv.tv_sec) * 1e6);
  setsockopt(fd, SOL_SOCKET, SO_RCVTIMEO, &tv, sizeof(tv));
  setsockopt(fd, SOL_SOCKET, SO_SNDTIMEO, &tv, sizeof(tv));

  struct sockaddr_in addr{};
  addr.sin_family = AF_INET;
  addr.sin_port   = htons(port_);
  if (inet_pton(AF_INET, host_.c_str(), &addr.sin_addr) != 1) {
    struct hostent* he = gethostbyname(host_.c_str());
    if (!he) {
      std::fprintf(stderr, "[dashboard] resolve %s failed\n", host_.c_str());
      ::close(fd);
      return false;
    }
    std::memcpy(&addr.sin_addr, he->h_addr, static_cast<size_t>(he->h_length));
  }
  if (::connect(fd, reinterpret_cast<struct sockaddr*>(&addr), sizeof(addr)) < 0) {
    std::fprintf(stderr, "[dashboard] connect(%s:%d) failed: %s\n",
                 host_.c_str(), port_, std::strerror(errno));
    ::close(fd);
    return false;
  }

  // drain welcome banner
  char buf[4096];
  ssize_t n = ::recv(fd, buf, sizeof(buf) - 1, 0);
  (void)n;

  sock_fd_ = fd;
  return true;
}

void DashboardClient::close() {
  if (sock_fd_ >= 0) {
    ::close(sock_fd_);
    sock_fd_ = -1;
  }
}

std::string DashboardClient::ask(const std::string& cmd) {
  if (sock_fd_ < 0 && !connect()) return "";
  std::string line = cmd + "\n";
  if (::send(sock_fd_, line.data(), line.size(), 0) < 0) {
    std::fprintf(stderr, "[dashboard] send(%s) failed: %s\n",
                 cmd.c_str(), std::strerror(errno));
    close();
    return "";
  }
  std::string resp;
  char buf[4096];
  while (resp.find('\n') == std::string::npos) {
    ssize_t n = ::recv(sock_fd_, buf, sizeof(buf), 0);
    if (n <= 0) break;
    resp.append(buf, buf + n);
  }
  // trim trailing \n
  while (!resp.empty() && (resp.back() == '\n' || resp.back() == '\r'))
    resp.pop_back();
  return resp;
}

int DashboardClient::get_robot_mode() {
  std::string resp = ask("robotmode");  // "Robotmode: RUNNING"
  auto colon = resp.rfind(':');
  std::string name = (colon == std::string::npos) ? resp
                     : resp.substr(colon + 1);
  // strip whitespace
  name.erase(0, name.find_first_not_of(" \t"));
  name.erase(name.find_last_not_of(" \t") + 1);
  for (int i = 0; i <= 7; ++i) {
    if (name == MODE_NAMES[i]) return i;
  }
  return -1;
}

std::string DashboardClient::get_safety_mode() { return ask("safetymode"); }

bool DashboardClient::is_in_remote_control() {
  std::string r = ask("is in remote control");
  std::transform(r.begin(), r.end(), r.begin(),
                 [](unsigned char c){ return std::tolower(c); });
  return r.rfind("true", 0) == 0;
}

std::string DashboardClient::power_on()      { return ask("power on"); }
std::string DashboardClient::power_off()     { return ask("power off"); }
std::string DashboardClient::brake_release() { return ask("brake release"); }
std::string DashboardClient::stop_program()  { return ask("stop"); }

bool DashboardClient::wait_for_mode(int target, double timeout_sec,
                                    double poll_sec) {
  auto deadline = std::chrono::steady_clock::now()
                + std::chrono::duration_cast<std::chrono::milliseconds>(
                      std::chrono::duration<double>(timeout_sec));
  int last = -2;
  while (std::chrono::steady_clock::now() < deadline) {
    int m = get_robot_mode();
    if (m != last) {
      std::fprintf(stderr, "[dashboard] robot_mode = %d (%s)\n",
                   m, robot_mode_name(m));
      last = m;
    }
    if (m == target) return true;
    std::this_thread::sleep_for(std::chrono::duration<double>(poll_sec));
  }
  std::fprintf(stderr, "[dashboard] wait_for_mode(%d) timed out (last %d=%s)\n",
               target, last, robot_mode_name(last));
  return false;
}

bool DashboardClient::power_up_sequence(double timeout_sec) {
  if (!connect()) return false;
  if (!is_in_remote_control()) {
    std::fprintf(stderr,
      "[dashboard] WARNING: robot NOT in Remote Control — power-cycle may fail.\n");
  }
  int mode = get_robot_mode();
  std::fprintf(stderr, "[dashboard] current mode = %d (%s)\n",
               mode, robot_mode_name(mode));
  if (mode == RM_RUNNING) {
    std::fprintf(stderr, "[dashboard] already RUNNING — skipping power-up\n");
    return true;
  }
  std::fprintf(stderr, "[dashboard] power on …\n");
  power_on();
  if (!wait_for_mode(RM_IDLE, timeout_sec)) return false;
  std::fprintf(stderr, "[dashboard] brake release …\n");
  brake_release();
  if (!wait_for_mode(RM_RUNNING, timeout_sec)) return false;
  std::fprintf(stderr, "[dashboard] power-up complete (RUNNING)\n");
  return true;
}

bool DashboardClient::power_down_sequence(double timeout_sec) {
  if (!connect()) return false;
  std::fprintf(stderr, "[dashboard] stop program …\n");
  stop_program();
  std::this_thread::sleep_for(std::chrono::milliseconds(500));
  std::fprintf(stderr, "[dashboard] power off …\n");
  power_off();
  wait_for_mode(RM_POWER_OFF, timeout_sec);
  std::fprintf(stderr, "[dashboard] power-down complete\n");
  return true;
}

}  // namespace ur10e_teleop_control_ff_cpp
